src/show.jl

```diff
@@ -16,35 +38,115 @@ function _string(x::MagOrRef)
     return read(out_rd, String)
 end
 
-function Base.show(io::IO, x::MagOrRef)
-    if isdefined(Main, :IJulia) && Main.IJulia.inited
-        print(io, Float64(x))
-    else
-        print(io, _string(x))
-    end
+function Base.string(
+    x::MagOrRef;
+    digits::Integer = digits_prec(30),
+    remove_trailing_zeros::Bool = true,
+)
+    cstr = ccall(@libflint(arf_get_str), Ptr{UInt8}, (Ref{arf_struct}, Int), Arf(x), digits)
+    str = unsafe_string(cstr)
+    ccall(@libflint(flint_free), Nothing, (Ptr{UInt8},), cstr)
+
+    return remove_trailing_zeros ? _remove_trailing_zeros(str) : str
+end
+
+function Base.string(
+    x::ArfOrRef;
+    digits::Integer = digits_prec(precision(x)),
+    remove_trailing_zeros::Bool = true,
+)
+    cstr = ccall(@libflint(arf_get_str), Ptr{UInt8}, (Ref{arf_struct}, Int), x, digits)
+    str = unsafe_string(cstr)
+    ccall(@libflint(flint_free), Nothing, (Ptr{UInt8},), cstr)
+
+    return remove_trailing_zeros ? _remove_trailing_zeros(str) : str
 end
 
-Base.show(io::IO, x::ArfOrRef) = print(io, BigFloat(x))
+function Base.string(
+    x::ArbOrRef;
+    digits::Integer = digits_prec(precision(x)),
+    more::Bool = false,
+    no_radius::Bool = false,
+    condense::Integer = 0,
+    unicode::Bool = false,
+    remove_trailing_zeros::Bool = !no_radius,
+)
+    flag = convert(UInt, more + 2no_radius + 16condense)
 
-function Base.show(io::IO, x::ArbOrRef)
     cstr = ccall(
         @libflint(arb_get_str),
         Ptr{UInt8},
         (Ref{arb_struct}, Int, UInt),
         x,
-        digits_prec(precision(x)),
-        UInt(0),
+        digits,
+        flag,
     )
-    print(io, unsafe_string(cstr))
+    str = unsafe_string(cstr)
     ccall(@libflint(flint_free), Nothing, (Ptr{UInt8},), cstr)
+
+    if unicode
+        # Multiple patterns in same call requires Julia 1.7
+        str = replace(replace(str, "+/-" => "±"), "..." => "…")
+    end
+
+    if remove_trailing_zeros && !startswith(str, '[')
+        str = _remove_trailing_zeros(str)
+    end
+
+    return str
+end
+
+function Base.string(
+    z::AcbOrRef;
+    digits::Integer = digits_prec(precision(z)),
+    more::Bool = false,
+    no_radius::Bool = false,
+    condense::Integer = 0,
+    unicode::Bool = false,
+    remove_trailing_zeros::Bool = true,
+)
+    str = string(
+        realref(z);
+        digits,
+        more,
+        no_radius,
+        condense,
+        unicode,
+        remove_trailing_zeros,
+    )
+
+    if !iszero(imagref(z))
+        str *=
+            " + " *
+            string(
+                imagref(z);
+                digits,
+                more,
+                no_radius,
+                condense,
+                unicode,
+                remove_trailing_zeros,
+            ) *
+            "im"
+    end
```

Comment:
```suggestion
    kwargs = (
        :digits => digits,
        :more => more,
        :no_radius => no_radius,
        :condense => condense,
        :unicode => unicode,
        :remove_trailing_zeros => remove_trailing_zeros
    )

    str = string(realref(z); kwargs...)

    if !iszero(imagref(z))
        str *= " + " * string(imagref(z); kwargs...) * "im"
    end
    return str
```
seems a bit more readable for me